js/private/node-patches/src/fs.ts

```diff
@@ -549,73 +454,270 @@ export const patcher = (fs: any = _fs, roots: string[]) => {
-                                // this is a broken symlink
-                                // even though this broken symlink points outside of the root
-                                // we'll return it.
-                                // the alternative choice here is to omit it from the directory listing altogether
-                                // this would add complexity because readdir output would be different than readdir
-                                // withFileTypes unless readdir was changed to match. if readdir was changed to match
-                                // it's performance would be greatly impacted because we would always have to use the
-                                // withFileTypes version which is slower.
-                                return resolve(v)
-                            }
-                            // transient fs related error. busy etc.
-                            return reject(err)
+                            throw err
                         }
-
-                        // add all stat is methods to Dirent instances with their result.
-                        v.isSymbolicLink = () =>
-                            origLstatSync(target).isSymbolicLink
                         patchDirent(v, stat)
                         resolve(v)
-                    }
-                )
+                    })
+                })
             })
         })
     }
 
-    function handleDirentSync(p: string, v: Dirent | null) {
+    function handleDirentSync(p: string, v: Dirent | null): void {
         if (v && v.isSymbolicLink) {
             if (v.isSymbolicLink()) {
-                // any errors thrown here are valid. things like transient fs errors
-                const target = path.resolve(
-                    p,
-                    origReadlinkSync(path.join(p, v.name))
-                )
-                if (isEscape(path.join(p, v.name), target)) {
-                    // Dirent exposes file type so if we want to hide that this is a link
-                    // we need to find out if it's a file or directory.
-                    v.isSymbolicLink = () =>
-                        origLstatSync(target).isSymbolicLink
-                    const stat: Stats | any = fs.statSync(target)
-                    // add all stat is methods to Dirent instances with their result.
+                const f = path.resolve(p, v.name)
+                if (f == guardedReadLinkSync(f)) {
+                    // There are no hops so we should hide the fact that the file is a symlink
+                    v.isSymbolicLink = () => false
+                    const stat = fs.statSync(origRealpathSync(f))
                     patchDirent(v, stat)
                 }
             }
         }
     }
+
+    function nextHop(loc: string, cb: (next: string | false) => void): void {
+        let nested = []
+        const oneHop = (maybe, cb: (next: string | false) => void) => {
```

Comment:
this `cb` is shadowing the one from the outer function, maybe we should have a lint for that?

Reply:
in this case it is intentional since it is the same cb.
```
const oneHop = (maybe, cb: (next: string | false) => void) => {...}
```
is followed by
```
oneHop(loc, cb)
```
which just passed the `cb` passed to the function forward